File: mobile-app/src/core/DIContainer.ts
/**
 * Dependency Injection Container
 * 
 * A lightweight DI container for managing service dependencies.
 * Supports singleton and transient service lifetimes.
 */

type ServiceFactory<T> = (container: DIContainer) => T;
type ServiceLifetime = 'singleton' | 'transient';

interface ServiceRegistration<T> {
  factory: ServiceFactory<T>;
  lifetime: ServiceLifetime;
  instance?: T;
}

export class DIContainer {
  private services = new Map<string, ServiceRegistration<any>>();
  private resolving = new Set<string>();

  /**
   * Register a service with the container
   * @param key - Unique identifier for the service
   * @param factory - Factory function to create the service
   * @param lifetime - Service lifetime (singleton or transient)
   */
  register<T>(
    key: string,
    factory: ServiceFactory<T>,
    lifetime: ServiceLifetime = 'singleton'
  ): void {
    if (this.services.has(key)) {
      throw new Error(`Service '${key}' is already registered`);
    }

    this.services.set(key, { factory, lifetime });
  }

  /**
   * Resolve a service from the container
   * @param key - Unique identifier for the service
   * @returns The resolved service instance
   */
  resolve<T>(key: string): T {
    const registration = this.services.get(key);

    if (!registration) {
      throw new Error(`Service '${key}' is not registered`);
    }

    // Check for circular dependencies
    if (this.resolving.has(key)) {
      throw new Error(`Circular dependency detected for service '${key}'`);
    }

    // Return existing singleton instance
    if (registration.lifetime === 'singleton' && registration.instance) {
      return registration.instance as T;
    }

    // Create new instance
    this.resolving.add(key);
    try {
      const instance = registration.factory(this);

      // Cache singleton instance
      if (registration.lifetime === 'singleton') {
        registration.instance = instance;
      }

      return instance;
    } finally {
      this.resolving.delete(key);
    }
  }

  /**
   * Check if a service is registered
   * @param key - Unique identifier for the service
   */
  has(key: string): boolean {
    return this.services.has(key);
  }

  /**
   * Clear all registered services and instances
   */
  clear(): void {
    this.services.clear();
    this.resolving.clear();
  }

  /**
   * Get all registered service keys
   */
  getRegisteredKeys(): string[] {
    return Array.from(this.services.keys());
  }
}

// Global container instance
export const container = new DIContainer();

File: mobile-app/ios/Core/DIContainer.swift
//
//  DIContainer.swift
//  ShadowShuttleTemp
//
//  Dependency Injection Container for iOS
//  Manages service dependencies with singleton and transient lifetimes
//

import Foundation

/// Service lifetime options
enum ServiceLifetime {
    case singleton
    case transient
}

/// Service registration information
private struct ServiceRegistration {
    let factory: (DIContainer) -> Any
    let lifetime: ServiceLifetime
    var instance: Any?
}

/// Dependency Injection Container
class DIContainer {
    private var services: [String: ServiceRegistration] = [:]
    private var resolving: Set<String> = []
    private let lock = NSLock()
    
    /// Register a service with the container
    /// - Parameters:
    ///   - type: The service type to register
    ///   - lifetime: Service lifetime (singleton or transient)
    ///   - factory: Factory closure to create the service
    func register<T>(_ type: T.Type, lifetime: ServiceLifetime = .singleton, factory: @escaping (DIContainer) -> T) {
        lock.lock()
        defer { lock.unlock() }
        
        let key = String(describing: type)
        
        guard services[key] == nil else {
            fatalError("Service '\(key)' is already registered")
        }
        
        services[key] = ServiceRegistration(
            factory: factory,
            lifetime: lifetime,
            instance: nil
        )
    }
    
    /// Resolve a service from the container
    /// - Parameter type: The service type to resolve
    /// - Returns: The resolved service instance
    func resolve<T>(_ type: T.Type) -> T {
        lock.lock()
        defer { lock.unlock() }
        
        let key = String(describing: type)
        
        guard var registration = services[key] else {
            fatalError("Service '\(key)' is not registered")
        }
        
        // Check for circular dependencies
        guard !resolving.contains(key) else {
            fatalError("Circular dependency detected for service '\(key)'")
        }
        
        // Return existing singleton instance
        if registration.lifetime == .singleton, let instance = registration.instance as? T {
            return instance
        }
        
        // Create new instance
        resolving.insert(key)
        defer { resolving.remove(key) }
        
        let instance = registration.factory(self) as! T
        
        // Cache singleton instance
        if registration.lifetime == .singleton {
            registration.instance = instance
            services[key] = registration
        }
        
        return instance
    }
    
    /// Check if a service is registered
    /// - Parameter type: The service type to check
    /// - Returns: True if the service is registered
    func has<T>(_ type: T.Type) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        
        let key = String(describing: type)
        return services[key] != nil
    }
    
    /// Clear all registered services and instances
    func clear() {
        lock.lock()
        defer { lock.unlock() }
        
        services.removeAll()
        resolving.removeAll()
    }
    
    /// Get all registered service keys
    /// - Returns: Array of registered service type names
    func getRegisteredKeys() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        
        return Array(services.keys)
    }
}

// Global container instance
let container = DIContainer()
